src/v/cloud_storage/remote_segment.cc

```diff
@@ -909,6 +909,19 @@ ss::future<> remote_segment::run_hydrate_bg() {
         }
     }
 
+    // If any new download requests got queued up while we were downloading
+    // chunks before the gate closed, cancel them so that the gate close does
+    // not get stuck during segment stop.
+    if (!_chunk_waiters.empty() && _gate.is_closed()) {
+        vlog(
+          _ctxlog.debug,
+          "Cancelling {} pending chunk downloads during segment stop",
+          _chunk_waiters.size());
+        for (auto& w : _chunk_waiters) {
+            w.promise.set_exception(ss::gate_closed_exception{});
+        }
```

Comment:
I think it makes sense, to handle any pending waiters after the gate close future resolves, I added something similar in the chunk API in this PR https://github.com/redpanda-data/redpanda/pull/18278/files#diff-989efee87e2ab784066d6a3b74db6e93214254d933c8dde8eae49ed6e0ca2901R104, I will look into adding a similar method to remote segment.

Reply:
> Just a thought, would it makes sense to call set_waiter_errors() in remote_segment::stop() to handle any stragglers? Or would that end up hanging because the in-flight hydrations wouldn't complete?

I tried this but it doesn't work, as you said the gate will never close because the task which added to chunk waiters holds it open.

I have been going through the code paths leading to this section and I don't think the code needs to change, it is correct as it is.

Consider the code flow related to chunk waiters modification 

(the code at the end of `cloud_storage::remote_segment::run_hydrate_bg` which cleans up waiters is referred to `cleanup`)

1. chunk waiters are added by either `cloud_storage::remote_segment::service_chunk_requests` or `cloud_storage::remote_segment::download_chunk`
2. the potential bug is that `cleanup` runs but does not handle all chunk waiters, IE some are added after `cleanup` runs or even concurrently.
3. `cleanup` runs when gate is closed, IE when the call to gate close is initiated, not when the future returned by `gate::close` resolves.

`cloud_storage::remote_segment::service_chunk_requests`: at the point when `cleanup` runs, we cannot have a concurrent execution of this method `sevice_chunk_requests` - this is part of the loop we just exited, and we `co_await` for the completion of `cloud_storage::remote_segment::service_chunk_requests` there as the last thing in the iteration. These two code sections are run serially.

`cloud_storage::remote_segment::download_chunk`: the first thing we do is try to enter the gate. At this point if the gate is closed, i.e. just the initial call to `gate::close` which sets the optional inside the gate, not the future being resolved; we will not be able to enter and add the waiter. If we do enter, which means the gate was open, from the gate entry to the point where we add the waiter there is no async call, so it is not possible that we enter the gate but somehow get pre-empted before adding the waiter, and add the waiter later on after `cleanup` runs.

 So it looks like we cannot add new waiters after or during `cleanup`, this code should handle all pending waiters.

@nvartolomei what do you think?

Reply:
Stared at this a bit and I think I agree that it's currently safe.

I think the general concern is around `service_chunk_requests()`, which has a scheduling point between the gate hold and the addition to `_chunk_waiters`. If the sequence was some:

- T1: begin call to `service_chunk_requests()`. Start holding the gate
- T1: swap out `_chunk_waiters` into `requests` for processing
- T2: call stop on the segment, close the gate 
- TN: background loop exits, collects the remainder of `_chunk_waiters`
- T1: process `requests`, some fail and are added to `_chunk_waiters`

perhaps that is what Nicolae's concerned about? In reality, TN is always going to come after the last step in T1 since they run from the same background fiber. Maybe that's useful context to include somewhere, or make the interface clearer e.g. by returning the failed chunk waiters in `service_chunk_requests()` and making the background loop explicitly handle that case